ffi_internals/src/impl_internals/fn_ffi.rs

```diff
@@ -134,28 +257,72 @@ impl FnFFI {
 
         let ffi_fn_name = self.ffi_fn_name(module_name);
         let native_fn_name = &self.fn_name;
-        let native_call = if self.has_receiver {
-            quote!(data.#native_fn_name)
+        let native_call = if self.receiver == FnReceiver::None {
+            if type_name.is_some() {
+                quote!(#type_name::#native_fn_name)
+            } else {
+                quote!(#native_fn_name)
+            }
         } else {
-            quote!(#type_name::#native_fn_name)
+            quote!(data.#native_fn_name)
         };
         let return_type = self
             .return_type
             .as_ref()
-            .map(|r| r.ffi_type(None, &Context::Return));
+            .map(|r| r.ffi_type(None, Context::Return));
         let call_and_return = if let Some(r) = &self.return_type {
             let assignment = quote!(let return_value = #native_call(#calling_args););
-            let mut return_conversion = if r.result {
-                let native_type = r.owned_native_type();
-                quote!(let return_value = ffi_common::ffi_core::try_or_set_error!(return_value, <#native_type>::default());)
+            let return_conversion = if r.is_result {
+                match &r.native_type {
+                    TypeIdentifier::Boxed(_)
+                    | TypeIdentifier::String
+                    | TypeIdentifier::DateTime
+                        if !r.is_vec =>
```

Comment:
It just occurred to me that `is_vec` isn't a function, which means it must be an attribute. ...which makes me wonder if whatever type `r` is could/should be an enum? Just thinking that introspection via a method isn't usually conventionally Rusty. Again, NBD, but just thought I'd mention it in case it makes anything nicer/easier for you.

Reply:
Hmm that's a really good question... This is the struct we have right now:

```
/// Describes the supported language-level generic wrappers around a `NativeType`, so that we can
/// expose an `Option<Foo>` or even a `Result<Vec<Foo>>`.
///
/// It's worth noting that these are only supported one level deep; we won't be able to expose a
/// `Vec<Vec<Foo>>` without making some larger improvements to the way we parse types.
///
#[derive(Debug)]
#[allow(clippy::struct_excessive_bools)]
pub struct TypeFFI {
    /// The underlying type being exposed.
    ///
    pub native_type: TypeIdentifier,
    /// True if `native_type` is wrapped in an `Option`, otherwise false.
    ///
    pub is_option: bool,
    /// True if `native_type` is the type of the elements in a `Vec` or slice, otherwise false.
    ///
    pub is_vec: bool,
    /// True if `native_type` is the type of the `Success` variant of a `Result`, otherwise false.
    ///
    pub is_result: bool,
    /// True if `native_type` is wrapped in a `Cow`, otherwise false.
    ///
    pub is_cow: bool,
    /// True if we're dealing with a borrowed reference to `native_type`./
    ///
    pub is_borrow: bool,
}
```

Right now, we only support particular configurations of these bools and don't handle them recursively (`Option<Vec<T>>` is fine, but something like `Vec<Result<Vec<T>>>` or `Option<Vec<Vec<Vec<T>>>>` is just going to choke). So yes, I think we could express what we support in an enum... But I think as we encounter more complex types and therefore need more complete language support, we're going to end up using https://docs.rs/syn/0.15.44/syn/enum.Type.html (or something very similar) a lot more heavily instead of just extracting bits of type information from it.

Which I suppose (after talking myself through this comment for 10 minutes) means I'd rather not refactor `TypeFFI` into an enum right now because I think it ought to be replaced by pushing `syn::Type` deeper into `ffi_internals` when we have a need for supporting more complex types.